teerex/src/lib.rs

```diff
@@ -446,6 +481,45 @@ impl<T: Config> Pallet<T> {
 			Err(<Error<T>>::RemoteAttestationTooOld.into())
 		}
 	}
+
+	fn check_block_and_get_latest_header(
+		shard_id: ShardIdentifier,
+		block_header: Header,
+		sender: &T::AccountId,
+		sender_index: u64,
+		latest_block_header: Header,
+	) -> Header {
+		log::error!("Importing new block, checking hash ...");
+		// Confirm that the parent hash is the hash of the previous block.
+		// Block number 1 does not have a previous block, hence skip checking there.
+		if latest_block_header.hash() == block_header.parent_hash || block_header.block_number == 1
+		{
+			log::error!("Hash correct.");
+			Self::confirm_sidechain_block(shard_id, block_header, &sender, sender_index);
+			return block_header
+		} else {
+			log::error!("Error: Parent hash of proposed block not correct!");
+		}
+		latest_block_header
+	}
+
+	fn confirm_sidechain_block(
+		shard_id: ShardIdentifier,
+		block_header: Header,
+		sender: &T::AccountId,
+		sender_index: u64,
+	) {
+		log::error!("Confirming block: {:?}", block_header.block_number);
+		<LatestSidechainBlockHeader<T>>::put(block_header);
+		<WorkerForShard<T>>::insert(shard_id, sender_index);
+		let block_hash = block_header.block_data_hash;
+		log::debug!(
+			"Proposed sidechain block confirmed with shard {:?}, block hash {:?}",
+			shard_id,
+			block_hash
+		);
+		Self::deposit_event(Event::ProposedSidechainBlock(sender.clone(), block_hash));
```

Comment:
How are sidechain blocks finalized in this way? Are they included in the parentchain block somehow or not? Does an event mean we have to subscribe to these events to get the information, or can we also get the information from a parentchain block?

Reply:
I have quickly spoken with @gaudenzkessler regarding this. I currently believe it is enough to just store the latest one. The worker shall find out on its own if it is not on the canonical chain, and fetch the correct blocks from another worker.

Reply:
That means however, that we roll-back a lot of sidechain blocks in case of a fork. If we only know the last finalized block, we don't know the exact block where the fork happened, and we need to roll-back everything until the last known finalized block. Which will be in the order of 40 sidechain blocks. E.g. if the fork happens at sidechain block 38 and we did know all the finalized blocks, we only roll back 2, not 40.

Reply:
The worker picks up every extrinsic that is included in a parentchain block, in this case that would be the `confirm_proposed_sidechain_block(origin: OriginFor<T>, shard_id: ShardIdentifier, block_header: Header)` extrinsic. So the worker itself knows the if an extrinsic with a specific header has been successful or not. It does not need the parentchain state for this.

Or am I missing something?

Reply:
But it needs to know if a sidechain block was finalized, not if an extrinsic was successful or not? An extrinsic can be successful, but still be on a sidechain block that will not be finalized.

Reply:
Ah, I see. So it's also successful if it's added into the queue. Hm. Then that doesn't work, my bad.

But lets say: The xt is successful if, and only if, the header gets added to the queue or is directly confirmed. And a header gets removed from the queue once it's been confirmed.

So depending on how we solve #48, can't we deduct the finalized headers by looking up the queue at the current parentchain block state, take all headers that have been successfully included in the block with `confirm_proposed_sidechain_block` minus the ones left in queue?

Reply:
Maybe we need to have a separate meeting for this, I think it's too much to discuss in the comments here. I believe the concept of `confirm_proposed_sidechain_block` will need to change or go away in the future, because it's not compatible with the finalization mechanism. When a sidechain block is proposed, even if it's perfectly valid from a worker perspective, it does not mean that it gets finalized by the parentchain.

Reply:
We decided to keep the latest block in order to keep the chain state small